plugins/module_utils/identity/keycloak/keycloak.py

```diff
@@ -207,24 +208,37 @@ def is_struct_included(struct1, struct2, exclude=None):
             Return True if all element of dict 1 are present in dict 2, return false otherwise.
     """
     if isinstance(struct1, list) and isinstance(struct2, list):
+        if len(struct1) == 0 and len(struct2) == 0:
+            return True
+        found = False
         for item1 in struct1:
-            if isinstance(item1, (list, dict)):
+            found = False
+            if isinstance(item1, list) or isinstance(item1, dict):
+                found1 = False
                 for item2 in struct2:
-                    if not is_struct_included(item1, item2, exclude):
-                        return False
+                    if is_struct_included(item1, item2, exclude):
+                        found1 = True
+                if found1:
+                    found = True
             else:
                 if item1 not in struct2:
                     return False
-        return True
+                else:
+                    found = True
+            if not found:
+                return False
+        return found
     elif isinstance(struct1, dict) and isinstance(struct2, dict):
+        if len(struct1) == 0 and len(struct2) == 0:
+            return True
         try:
             for key in struct1:
                 if not (exclude and key in exclude):
                     if not is_struct_included(struct1[key], struct2[key], exclude):
                         return False
-            return True
         except KeyError:
             return False
+        return True
```

Comment:
I'm very confused over the above changes to `is_struct_included()`. They seem to make the function more complex without changing behavior? (The only things I would accept are the length checks, but there I would not explicitly compute `len(...)` but instead use `not ...` to check whether the dict/list is empty.)

Reply:
I ported the code form our ansible fork we use for many years to manage Keycloak. I cannot remember why i coded it that way at that time but what I know is it make the comparison work with composite role. I had to do this change to be able to go deeper in the analysis of the two structures to be compared.
The original code would not compare the composites in a good way and will do the change even if no changes is needed.
If you execute the unit tests I added with the old code il will fail because sub options are not compared efficiently.
I can try to refactor the code if you really think it is needed.

Reply:
Ah, I see, there is a bug in this function. This should be fixed in a separate PR, though, so it can be backported to community.general 6.x.y as well.